pallets/gear/src/tests.rs

```diff
@@ -81,6 +82,158 @@ use utils::*;
 
 type Gas = <<Test as Config>::GasProvider as common::GasProvider>::GasTree;
 
+#[test]
+fn err_reply_comes_with_value() {
+    init_logger();
+    new_test_ext().execute_with(|| {
+        const VALUE: u128 = 10_000_000_000_000;
+
+        let (_init_mid, pid) = init_constructor(Scheme::empty());
+
+        // Case #1.
+        // If reply-able message quits with error, value is attached to the reply.
+        let user_balance = Balances::free_balance(USER_1);
+        assert_eq!(Balances::free_balance(pid.cast::<AccountId>()), get_ed());
+
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            vec![],
+            0,
+            VALUE,
+            false,
+        ));
+        assert_balance(USER_1, user_balance - VALUE, VALUE);
+        assert_eq!(Balances::free_balance(pid.cast::<AccountId>()), get_ed());
+
+        run_to_next_block(None);
+
+        assert_balance(USER_1, user_balance, 0u8);
+        assert_eq!(Balances::free_balance(pid.cast::<AccountId>()), get_ed());
+
+        let err_reply = maybe_last_message(USER_1).expect("Message should be");
+
+        assert_eq!(
+            err_reply.reply_code().expect("must be"),
+            ReplyCode::Error(ErrorReplyReason::Execution(
+                SimpleExecutionError::RanOutOfGas
+            ))
+        );
+
+        assert_eq!(err_reply.value(), VALUE);
+
+        // Cases #2-3.
+        // If non-reply-able message quits with error, value is kept by program.
+        //
+        // Case #2: success reply quits with error.
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_2),
+            pid,
+            Calls::builder()
+                .send(USER_1.into_origin().0, b"Hello, world!".to_vec())
+                .encode(),
+            BlockGasLimitOf::<Test>::get(),
+            0,
+            false,
+        ));
+
+        run_to_next_block(None);
+
+        let mail = get_last_mail(USER_1);
+
+        assert_ok!(Gear::send_reply(
+            RuntimeOrigin::signed(USER_1),
+            mail.id(),
+            vec![],
+            0,
+            VALUE,
+            false,
+        ));
+
+        assert_balance(USER_1, user_balance - VALUE, VALUE);
+
+        run_to_next_block(None);
+
+        let pid_balance = Balances::free_balance(pid.cast::<AccountId>());
+        assert_eq!(pid_balance, get_ed() + VALUE);
+
+        assert_balance(USER_1, user_balance - VALUE, 0u8);
+        assert_balance(pid, pid_balance, 0u8);
```

Comment:
At first I thought that it was implementation error and a bug. But from tests I see that it was done intentionally.
So in case non-reply-able message with value fails you still transfer the value.
What benefits are intended to gain with that?

That seems to be inconsistent with a general behavior of the system - in case of error all important state transitions must be reverted.

Reply:
discussed

I'm still against these changes, but it doesn't bring that much changes to the development of smart contracts, so I'm ok with introducing a new invariant